platform/logger.h

```diff
@@ -0,0 +1,150 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** \addtogroup platform */
+/** @{*/
+/**
+ * \defgroup platform_log Logger functions
+ * @{
+ */
+
+#ifndef LOGGER_INTERNAL_H
+#define LOGGER_INTERNAL_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include "platform/mbed_preprocessor.h"
+
+#ifndef MBED_CONF_MAX_LOG_STR_SIZE
+#define MBED_CONF_MAX_LOG_STR_SIZE     128
+#endif
+
+#ifndef MBED_CONF_LOG_MAX_BUFFER_SIZE
+#define MBED_CONF_LOG_MAX_BUFFER_SIZE  1024
+#endif
+
+#if defined (MBED_ID_BASED_TRACING)
+#define LOG_DATA_TYPE_                 uint32_t
+#define LOG_SINGLE_STR_SIZE_           4
+#else
+#define LOG_DATA_TYPE_                 char
+#define LOG_SINGLE_STR_SIZE_           MBED_CONF_MAX_LOG_STR_SIZE
+#endif
+
+#if defined(__ARMCC_VERSION)
+#define LOG_FILE_NAME_           __MODULE__
+#else
+#define LOG_FILE_NAME_           __BASE_FILE__
+#endif
+
+// Log-Level Strings
+#define LOG_GEN_                 "GEN "
+#define LOG_ERR_CRITICAL_        "CRT "
+#define LOG_ERR_                 "ERR "
+#define LOG_WARN_                "WARN"
+#define LOG_DEBUG_               "DBG "
+#define LOG_INFO_                "INFO"
+#define LOG_TRACE_               "TRAC"
+
+#define GET_LOG_STRING(ll)      ((ll > LOG_LEVEL_DEBUG) ? LOG_TRACE_ : \
+                                ((ll > LOG_LEVEL_INFO)  ? LOG_DEBUG_ : \
+                                ((ll > LOG_LEVEL_WARN)  ? LOG_INFO_  : \
+                                ((ll > LOG_LEVEL_ERR)   ? LOG_WARN_  : \
+                                ((ll > 1)   ? LOG_ERR_   : LOG_GEN_)))))
+
+#define GET_MODULE_ID_(a, b, c, d)   (375 + (a << 1) + (b << 2) + (c >> 2) + (d >> 1))
+#define GET_MODULE_ID(x)        (GET_MODULE_ID_(x[0], x[1], x[2], x[3]))
+
+#define SET_MODULE(x)           (GET_MODULE_ID(x) & 0xFF)
+#define SET_COUNTER(y)          ((y & 0xFF) << 8)
+#define SET_LINE_NUM(z)         ((z & 0xFFFF) << 16)
+#define TRACE_ID_(x,y,z)        (SET_MODULE(x) | SET_COUNTER(y) | SET_LINE_NUM(z))
+
+// Macros to log ID based data
+#define MBED_LOG_ID_4(...)                            log_buffer_id_data(__VA_ARGS__)
+// Data dumped in special section is : Unique ID, format string length and format string
+#if 0
```

Comment:
dead code?

Reply:
Dead code is for ID based tracing, will be enabled later. Needs linker file changes